src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,309 @@
+        const char *kvsep = "\x01";
+        const int kvsep_size = strlen(kvsep);
+        int extension_size = 0;
+        int i;
+        for (i = 0 ; i < rd_list_cnt(&state->extensions) ; i++) {
+                rd_strtup_t *extension = rd_list_elem(&state->extensions, i);
+                // kvpair         = key "=" value kvsep
+                extension_size += strlen(extension->name) + 1 // "="
+                        + strlen(extension->value) + kvsep_size;
+        }
+
+        // client-resp    = (gs2-header kvsep *kvpair kvsep) / kvsep
+        out->size = strlen(gs2_header) + kvsep_size
+                + strlen("auth=Bearer ") + strlen(state->token_value) + kvsep_size
+                + extension_size + kvsep_size;
+        out->ptr = rd_malloc(out->size+1);
+
+        char *buf = out->ptr;
+        int size_written = 0;
+        size_written += rd_snprintf(buf, out->size+1 - size_written,
+                "%s%sauth=Bearer %s%s", gs2_header, kvsep, state->token_value, kvsep);
+        buf = out->ptr + size_written;
+
+        for (i = 0 ; i < rd_list_cnt(&state->extensions) ; i++) {
+                rd_strtup_t *extension = rd_list_elem(&state->extensions, i);
+                size_written += rd_snprintf(buf, out->size+1 - size_written,
+                        "%s=%s%s", extension->name, extension->value, kvsep);
+                buf = out->ptr + size_written;
+        }
+
+        rd_snprintf(buf, out->size+1 - size_written, "%s", kvsep);
+
+        rd_rkb_dbg(rktrans->rktrans_rkb, SECURITY, "SASLOAUTHBEARER",
+                   "Built client first message");
+}
+
+
+
+/**
+ * @brief SASL OAUTHBEARER client state machine
+ * @returns -1 on failure (errstr set), else 0.
+ */
+static int rd_kafka_sasl_oauthbearer_fsm (rd_kafka_transport_t *rktrans,
+                                    const rd_chariov_t *in,
+                                    char *errstr, size_t errstr_size) {
+        static const char *state_names[] = {
+                "client-first-message",
+                "server-first-message",
+                "server-failure-message",
+        };
+        struct rd_kafka_sasl_oauthbearer_state *state = rktrans->rktrans_sasl.state;
+        rd_chariov_t out = RD_ZERO_INIT;
+        int r = -1;
+
+        rd_rkb_dbg(rktrans->rktrans_rkb, SECURITY, "SASLOAUTHBEARER",
+                   "SASL OAUTHBEARER client in state %s",
+                   state_names[state->state]);
+
+        switch (state->state)
+        {
+        case RD_KAFKA_SASL_OAUTHBEARER_STATE_SEND_CLIENT_FIRST_MESSAGE:
+                rd_dassert(!in); /* Not expecting any server-input */
+
+                rd_kafka_sasl_oauthbearer_build_client_first_message(rktrans, &out);
+                state->state = RD_KAFKA_SASL_OAUTHBEARER_STATE_RECEIVE_SERVER_FIRST_MESSAGE;
+                break;
+
+
+        case RD_KAFKA_SASL_OAUTHBEARER_STATE_RECEIVE_SERVER_FIRST_MESSAGE:
+                if (!in || in->ptr[0] == '\0') {
```

Comment:
`in` should always be non-NULL in this state.
Also make sure to check length of in before dereferncing to allow empty messages from the broker.
So..
```
  rd_assert(in);
   if (in->size == 0 || !*in->ptr) {
  ..
```

Reply:
👍 Fixed